### PR TITLE
Store gateway mmap removal: use a shared pool of buffers across all Decbuf instances

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

Previously, we created a separate pool of `bufio.Reader`s for each `DecbufFactory` (and we create a new factory for each index-header). This is unnecessary - we can share the same pool of buffers across all `Decbuf`s.

#### Which issue(s) this PR fixes or relates to

#3465 

#### Checklist

- [n/a] Tests updated
- [n/a] Documentation added
- [n/a] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/storegateway/indexheader/encoding/reader.go [hunk 1]
```diff
@@ -139,9 +146,9 @@ func (f *FileReader) Len() int {
 	return f.length - f.pos
 }
 
-// close closes the underlying resources used by this FileReader. This method
-// is unexported to ensure that all resource management is handled by DecbufFactory
-// which pools resources.
-func (f *FileReader) close() error {
+// Close closes the underlying resources used by this FileReader.
+func (f *FileReader) Close() error {
+	bufferPool.Put(f.buf)
```
Comment (colega) [inline pkg/storegateway/indexheader/encoding/reader.go:151]: I guess we need to reset the buffer state before returning it?

Comment (charleskorn) [reply pkg/storegateway/indexheader/encoding/reader.go:151]: We reset it when we retrieve it from the pool: `NewFileReader()` calls `Reset()`, which calls `ResetAt()`, which then resets the buffer: https://github.com/grafana/mimir/pull/3691/files#diff-2ae6b52cbe80bef6dd8daedc332df3984d2394f761480ff8aeb94cc3586888a8R66

Comment (pracucci) [reply pkg/storegateway/indexheader/encoding/reader.go:151]: I would add a comment here to explain it.

**Review by pracucci — APPROVED** [review]
LGTM, but let's wait for @colega follow up too.

**Review by colega — APPROVED** [review]

File: pkg/storegateway/indexheader/encoding/factory.go [hunk 2]
```diff
@@ -124,33 +114,12 @@ func (df *DecbufFactory) NewRawDecbuf() Decbuf {
...
+	reader, err := NewFileReader(f, 0, int(fileSize))
 	if err != nil {
 		return Decbuf{E: errors.Wrap(err, "file reader for decbuf")}
 	}
 
 	closeFile = false
 	return Decbuf{r: reader}
 }
-
-// Close cleans up any resources associated with the Decbuf
```
[the remark this answers is not in the document]
Comment (56quarters) [reply pkg/storegateway/indexheader/encoding/factory.go:138]: Alternatively for my file-handle work, the `FileReader` / `Decbuf` could be modified to keep a reference to the `DecbufFactory` and return any resources to it. That would keep their `.Close()` methods as the canonical way to clean them up. That feels a bit too clever to me though. Curious about what other people think.

[the remark this answers is not in the document]
Comment (charleskorn) [reply pkg/storegateway/indexheader/encoding/factory.go:138]: I'm happy to keep `DecbufFactory.Close()` - I'll revert that change.

File: pkg/storegateway/indexheader/encoding/reader.go (hunk 1 above)
Comment (charleskorn) [reply pkg/storegateway/indexheader/encoding/reader.go:155]: I've added a comment in 9e85f11.

**Review by 56quarters — APPROVED** [review]

**Pull request merged**